### PR TITLE
Improve multiple error messages

### LINKED ISSUE #12547: Packed structs can't contain arrays
### Zig Version

0.10.0-dev.3659+e5e6eb983

### Steps to Reproduce

```
matteo@lab:/tmp$ zig version
0.10.0-dev.3659+e5e6eb983
matteo@lab:/tmp$ cat test.zig
const std = @import("std");

const TestStruct = packed struct {
    pad: [512]u8,
    x: u64,
};

pub fn main() void {
    const asdf = TestStruct{.pad = [_]u8{0} ** 512, .x = 1};
    std.debug.print("{d}", .{asdf.x});
}
matteo@lab:/tmp$ zig build-exe test.zig
test.zig:4:5: error: packed structs cannot contain fields of type '[512]u8'
    pad: [512]u8,
    ^~~~~~~~~~~~
test.zig:4:5: note: type has no guaranteed in-memory representation
matteo@lab:/tmp$ zig build-exe -fstage1 test.zig
matteo@lab:/tmp$
```

### Expected Behavior

The example compiles without errors with both `-fstage1` and with the default settings.

### Actual Behavior

The compiler complains that `[512]u8` can't be in a packed struct because "it has no guaranteed in-memory representation". This doesn't really make sense to me, isn't an u8 array supposed to be just a fixed-size sequence of bytes?

When using the `-fstage1` the compiler doesn't complain.

### LINKED ISSUE #12829: Misleading error message for `@floatToInt`
### Zig Version

0.10.0-dev.3880+e2bb92b2e

### Steps to Reproduce

Call `@floatToInt` for a `nan` value, like this:
```
const std = @import("std");

pub fn main() void {
	var float: f32 = 0;
	float /= 0.0;
    std.log.err("{}", .{float});
    std.log.err("{}", .{@floatToInt(u32, float)});
}
```

### Expected Behavior

The error message should either explicitely list the nan/infinity case:
`panic: integer part of floating point value out of bounds or not finite`
Or there should be seperate error message for nan values:
`panic: cannot cast nan value to int`

I'd personally prefer the second one, because it gives me more data that I can use for debugging, but the first one would also be fine.

### Actual Behavior

`panic: integer part of floating point value out of bounds`

To me this error message implies that the float has an integer part and therefor is finite.

### LINKED ISSUE #13832: compile error should suggest to use error set instead of "expected type 'type'"
### Zig Version

0.11.0-dev.635+225ed65ed2

### Steps to Reproduce and Observed Output

```zig
// t.zig: zig run t.zig
const std = @import("std");

pub fn main() void {
    somefunction() catch {};
}

fn somefunction() error.name!void {
    return error.name;
}
```

Output making some new people believe that something went wrong with the coercion to the error set:
```
min.zig:7:19: error: expected type 'type', found 'error{name}'
fn somefunction() error.name!void {
                  ^~~~~~~~~~
```

### Expected Output

```
t.zig:7:19: error: expected error set, found 'error.name'
fn somefunction() error.name!void {
                  ^~~~~~~~~~
```

### LINKED ISSUE #14238: Add a hint when trying to print at compile time
### Zig Version

0.11.0-dev.1199+1878bdfbb1

### Steps to Reproduce and Observed Output

While debugging, I added something along those lines:

```
std.debug.print("error: {}", .{err});
```

In a comptime context.

The error I got:

```
/usr/lib/zig/std/mem.zig:205:17: error: unable to evaluate comptime expression
        dest[i] = s;
        ~~~~~~~~^~~
/usr/lib/zig/std/mem.zig:205:13: note: operation is runtime due to this operand
        dest[i] = s;
        ~~~~^~~
/usr/lib/zig/std/io/fixed_buffer_stream.zig:70:21: note: called from here
            mem.copy(u8, self.buffer[self.pos .. self.pos + n], bytes[0..n]);
            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/std/io/writer.zig:17:27: note: called from here
            return writeFn(self.context, bytes);
                   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/std/io/writer.zig:23:40: note: called from here
                index += try self.write(bytes[index..]);
                             ~~~~~~~~~~^~~~~~~~~~~~~~~~
/usr/lib/zig/std/fmt.zig:121:32: note: called from here
            try writer.writeAll(fmt[start_index..end_index]);
                ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/std/fmt.zig:1954:15: note: called from here
    try format(fbs.writer(), fmt, args);
        ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/home/kuon/pbg/zkfs-sdk/js/src/common/log.zig:16:31: note: called from here
    var buf = std.fmt.bufPrint(&log_buf, format ++ "\r\n", args) catch {
              ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/std/log.zig:125:22: note: called from here
    std.options.logFn(message_level, scope, format, args);
    ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/std/log.zig:203:16: note: called from here
            log(.debug, scope, format, args);
            ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/client.zig:17:14: note: called from here
        debug("error: {}", .{err});
        ~~~~~^~~~~~~~~~~~~~~~~~~~~
src/client.zig:24:24: note: called from here
```

It took me a while to realize what I was doing wrong.

### Expected Output

A hint along those lines would be very helpful to fix those errors:

```
Cannot print to std outputs at compile time, use @compileLog or @compileError instead.
```

### LINKED ISSUE #14334: Error message assumes `struct` is a tuple when some members miss type
### Zig Version

0.11.0-dev.1302+d813cef42

### Steps to Reproduce and Observed Output

Minimal example:
```zig
pub const Foo = struct {
	a: u32,
	b,
};
pub const Bar = struct {
	pub fn bar() void {}
	a: u32,
	b,
};
```
For both of these structs, because the second member could syntactically be a tuple member, the error messages are formulated under the assumption that the type were in fact a tuple.
Observed Output:
```
main.zig:2:2: error: tuple field has a name
 a: u32,
 ^
main.zig:6:6: error: tuple declarations cannot contain declarations
 pub fn bar() void {}
 ~~~~^~~~~~~~~~~~~~~~
main.zig:7:2: note: tuple field here
 a: u32,
 ^~~~~~
```
For the second error, the generated `note` is not even technically correct, as the error message just points at the first field, which is not even syntactically a valid tuple field.

### Expected Output

1. The error messages for `Foo` should mention both options, removing field types to make a tuple, as well as adding field types to make a valid non-tuple `struct`.
   - IMO `struct` should be mentioned first by default, since a usual code base has more non-tuple `struct`s than tuples.
     (Also, aesthetically, because tuples are generally less explicit, while Zig overall encourages explicitness.)
2. In the case of `Bar`, IMO the fact that the fields make the type invalid is more fundamental and pressing than the fact it has declarations.
   Having both errors would be okay, but if only one is reported, it should IMO be the other one.

### LINKED ISSUE #16019: Incorrect error message when assigning to a field in a struct definition
### Zig Version

0.11.0-dev.3395+1e7dcaa3a

### Steps to Reproduce and Observed Output

```zig
const S = struct {
    foo: u32,
    bar: u32,
    baz = 5,
};
```

Output:
```
~ % zig build-obj test.zig
test.zig:2:5: error: tuple field has a name
    foo: u32,
    ^~~
```

### Expected Output

Caret should point to `baz` and the message should be more clear (if the `foo` and `bar` fields aren't present, the error message is `use of undeclared identifier 'baz'`)

### LINKED ISSUE #16702: Canonicalize identifiers in error messages
### Zig Version

0.12.0-dev.5+fc6e57568

### Steps to Reproduce and Observed Output

```
$ echo 'const x = @"a\nb";' > test.zig
$ zig build-exe test.zig
test.zig:1:11: error: use of undeclared identifier 'a
                      b'
const x = @"a\nb";
          ^~~~~~~
```

### Expected Output

I'd prefer if the error message represented the identifier the same way it looks in Zig source:

```
test.zig:1:11: error: use of undeclared identifier @"a\nb"
const x = @"a\nb";
          ^~~~~~~
```

In other words, print identifiers the same way zig fmt would canonicalize them (#166). And if that requires @"...", then don't bother adding single quotes around it.

### PR DESCRIPTION
**Pull request by tau-dev**
Fixes #12547.
Fixes #12829.
Fixes #13832.
Fixes #14334.
Fixes #16019.
Fixes #16702.
Fixes #14238.

None of these are particularly related, I can pull them into separate PRs if necessary.

### CODE REVIEW

File: lib/std/zig/AstGen.zig [hunk 1]
```diff
@@ -13815,12 +13871,25 @@ fn lowerAstErrors(astgen: *AstGen) !void {
     const gpa = astgen.gpa;
     const parse_err = tree.errors[0];
 
-    var msg: std.ArrayListUnmanaged(u8) = .{};
-    defer msg.deinit(gpa);
-
     const token_starts = tree.tokens.items(.start);
     const token_tags = tree.tokens.items(.tag);
 
+    if (try @import("Parse.zig").findUnmatchedParen(astgen.gpa, token_tags)) |tok| {
```
Comment (gooncreeper) [inline lib/std/zig/AstGen.zig:13877]: Import should be at top of file

File: src/Sema.zig [hunk 2]
```diff
@@ -27010,8 +27014,8 @@ fn explainWhyTypeIsNotPacked(
         .ErrorSet,
         .AnyFrame,
         .Optional,
-        .Array,
         => try sema.errNote(src_loc, msg, "type has no guaranteed in-memory representation", .{}),
+        .Array => try sema.errNote(src_loc, msg, "packed fields are ordered according to machine endianness, array elements are not", .{}),
```
Comment (gooncreeper) [inline src/Sema.zig:27018]: Error message sounds more like a warning and should not explain how packed structs work.
A better error message would be something like "arrays are not allowed in packed structs"

File: test/compile_errors.zig [hunk 3]
```diff
@@ -38,6 +38,74 @@ pub fn addCases(ctx: *Cases, b: *std.Build) !void {
...
+
+    {
+        const case = ctx.obj("unmatched parentheses", b.graph.host);
+
+        case.addError(
+            \\export fn a() void {
+            \\}
+            \\}
+        , &[_][]const u8{
+            ":3:1: error: unmatched curly brace",
```
Comment (gooncreeper) [inline test/compile_errors.zig:73]: Two error messages for one problem seems bad
It would be better if this was a note

Comment (tau-dev) [reply test/compile_errors.zig:73]: Most of the time, the unmatched brace is the underlying error and the particular parse error coming out of that is just a symptom, so I'd definitely expect the unmatched-error first, not as a note on something else.

File: src/Sema.zig (hunk 2 above)
Comment (tau-dev) [reply src/Sema.zig:27018]: The reason for it not being allowed is rather subtle, and I think explaining it helps the user understand whether they want to solve the error by giving up packed-ness or array-ness. As far as I understand, the reason #12547 is still open is that the status quo error message saying only "this isn't allowed" is not useful to new users.

**Review by andrewrk — CHANGES_REQUESTED** [review]
I'm sorry but I do request these to be separate PRs. I went through the first 3 commits and saw things that I want to ask about in each one.

**Comment by tau-dev** [comment]
Split this into #20760, #20761, #20762, #20763 and #20764.

**Pull request closed without merging**